File: opencode-plugin/tests/acceptance_live.py
"""Day 4 acceptance check: run the plugin's hooks against a LIVE engine and print the evidence.

This is the leg the automated harnesses cannot cover — a real engine, real Postgres, real project
state. It needs no LLM key and no coding session: it drives the same hooks OpenCode calls and
prints the exact block the agent would receive, which is the artifact to paste into HANDOFF.md.

    python opencode-plugin/tests/acceptance_live.py --self-test   # fixture engine (proves the runner)
    python opencode-plugin/tests/acceptance_live.py               # $SCAFFOLD_ENGINE_URL or localhost:8000
    python opencode-plugin/tests/acceptance_live.py --url http://192.168.1.20:8000

Requires a live engine whose .env has DATABASE_URL + SCAFFOLD_DEFAULT_PROJECT_ID
(engine/README steps: pip install -r requirements.txt, apply schema, seed, uvicorn app.main:app).

    SCAFFOLD_ACCEPTANCE_REPORT=1 python opencode-plugin/tests/acceptance_live.py
also exercises report_change — that WRITES a `change_reported` event to the real project.
"""

import argparse
import os
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))

from engine_fixture import FixtureEngine, check, report, run_driver  # noqa: E402

DRIVER = Path(__file__).resolve().with_name("acceptance_live.ts")


def run(url: str, label: str) -> None:
    print(f"\n== {label}: {url} ==")
    driver, proc = run_driver(DRIVER, url)

    # The block is the artifact a human needs to see, injected or not (plus why, if not).
    print(proc.stdout.rstrip())
    if proc.stderr.strip():
        print(proc.stderr.rstrip())

    if driver is None:
        check(f"{label}: driver produced a result", False, f"exit {proc.returncode}")
        return

    check(f"{label}: engine answered and context was injected", driver["blockChars"] > 0, "; ".join(driver["warnings"]))
    check(f"{label}: injected block stayed bounded", driver["blockChars"] <= 2400, f"{driver['blockChars']} chars")
    check(f"{label}: no engine warnings while reachable", not driver["warnings"], "; ".join(driver["warnings"]))
    if driver["reportsAttempted"]:
        check(f"{label}: report_change round trip accepted", driver["reportsToasted"])


parser = argparse.ArgumentParser(description=__doc__)
parser.add_argument("--self-test", action="store_true", help="run against the fixture engine instead of a live one")
parser.add_argument("--url", default=os.environ.get("SCAFFOLD_ENGINE_URL", "http://localhost:8000"))
args = parser.parse_args()

if args.self_test:
    # Same code path as the live run, pointed at the fixture: proves the runner itself works.
    with FixtureEngine() as engine:
        run(engine.url, "self-test (fixture engine)")
        check("self-test: the fixture actually injected a block", len(engine.calls) > 0, str(engine.calls))
else:
    run(args.url, "live engine")
    print(
        "\nEvidence for docs/HANDOFF.md: paste the injected block above.\n"
        "For the Day 4 handoff requirement, run this on BOTH machines after Dev A commits —\n"
        "Dev B's block should already carry Dev A's change (decisions/contracts), with nobody restarting anything."
    )

sys.exit(report())

File: opencode-plugin/tests/acceptance_live.ts
/**
 * The evidence driver for `tests/acceptance_live.py`: what a real coding agent would actually
 * receive from a **live** Scaffold engine, printed verbatim so it can be pasted into HANDOFF.md.
 *
 * Not runnable on its own — the runner sets SCAFFOLD_ENGINE_URL and reads the `__RESULT__` line.
 * Set SCAFFOLD_ACCEPTANCE_REPORT=1 to also exercise the report_change round trip (that WRITES a
 * `change_reported` event into the real project, so it is opt-in).
 */

const engineUrl = process.env.SCAFFOLD_ENGINE_URL ?? "http://localhost:8000"
const shouldReport = process.env.SCAFFOLD_ACCEPTANCE_REPORT === "1"

const { ScaffoldPlugin } = await import(new URL("../.opencode/plugins/scaffold.ts", import.meta.url).href)

const logs: any[] = []
const toasts: any[] = []
const client = {
  app: { log: async ({ body }: any) => void logs.push(body) },
  tui: { showToast: async ({ body }: any) => void toasts.push(body) },
}
const cwd = process.cwd()
const hooks = await ScaffoldPlugin({
  client,
  project: { id: "acceptance-check" },
  directory: cwd,
  worktree: cwd,
  $: async () => {},
} as any)

// Exactly what OpenCode does at the start of a session.
await hooks.event!({ event: { type: "session.created", properties: { sessionID: "ses-acceptance" } } } as any)
const system = ["base system prompt"]
const baseEntries = system.length
await hooks["experimental.chat.system.transform"]!({ sessionID: "ses-acceptance" }, { system } as any)
// Only entries APPENDED by the plugin count — the base prompt is not an injected block.
const block = system.length > baseEntries ? system.slice(baseEntries).join("\n\n") : ""

console.log("----- injected block (verbatim) -----")
console.log(block || "(nothing injected)")
console.log("----- end injected block -----")

const warnings = logs.filter((l) => l.level === "warn")
for (const warning of warnings) console.error(`warning: ${warning.message}`)

let reported = false
if (shouldReport) {
  // A write to a clearly synthetic path, so nobody mistakes this for real work in the feed.
  await hooks["tool.execute.before"]!(
    { tool: "write", sessionID: "ses-acceptance", callID: "acceptance-write" },
    { args: { filePath: `${cwd}/acceptance-check.txt`, content: "scaffold acceptance check\n" } },
  )
  await hooks["tool.execute.after"]!(
    {
      tool: "write",
      sessionID: "ses-acceptance",
      callID: "acceptance-write",
      args: { filePath: `${cwd}/acceptance-check.txt`, content: "scaffold acceptance check\n" },
    },
    { title: "acceptance-check.txt", output: "", metadata: { filepath: `${cwd}/acceptance-check.txt`, exists: false } } as any,
  )
  reported = toasts.length > 0
}

console.log(
  `__RESULT__${JSON.stringify({
    engineUrl,
    blockChars: block.length,
    warnings: warnings.map((w) => w.message),
    reportsAttempted: shouldReport,
    reportsToasted: reported,
  })}`,
)
process.exit(0)
